src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const currentState = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(currentState, extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in currentState) {
+          delete currentState[key];
+        }
+        break;
+    }
+
+    stateHistory.push({ ...currentState });
```

Comment:
You are pushing the new state into 'stateHistory' after each action, which is correct, but you should make the copy of the state before applying the next action to avoid mutating the previous states in the 'stateHistory' array. This can be done by moving the copying operation to the beginning of the loop.